{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"name":"11_Multiple_Testing_Python.ipynb","provenance":[],"authorship_tag":"ABX9TyMpEELS79QwmEXDDwaekUtN"},"kernelspec":{"name":"python3","display_name":"Python 3"}},"cells":[{"cell_type":"markdown","metadata":{"id":"4BLWxIbNa-TI"},"source":["## MultiPy\r\n","\r\n","Testing multiple hypotheses simultaneously increases the number of false positive findings if the corresponding p-values are not corrected. While this multiple testing problem is well known, the classic and advanced correction methods are yet to be implemented into a coherent Python package. This package sets out to fill this gap by implementing methods for controlling the family-wise error rate (FWER) and the false discovery rate (FDR).\r\n","\r\n"]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"wHXAN8IrbHBt","executionInfo":{"status":"ok","timestamp":1610435322800,"user_tz":-60,"elapsed":6906,"user":{"displayName":"sebastian heucke","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GgU7r1bq5bIRNqaoCYdLPidtmcPCdoxFiu0JStl=s64","userId":"03230349622000046876"}},"outputId":"41770c84-7eeb-4fad-b929-916c11889ee4"},"source":["!pip install multipy"],"execution_count":1,"outputs":[{"output_type":"stream","text":["Collecting multipy\n","  Downloading https://files.pythonhosted.org/packages/b4/a5/15558a909f56a080ba3721bafd446a6074fe4615013168f378340e54c71c/multipy-0.16.tar.gz\n","Building wheels for collected packages: multipy\n","  Building wheel for multipy (setup.py) ... \u001b[?25l\u001b[?25hdone\n","  Created wheel for multipy: filename=multipy-0.16-cp36-none-any.whl size=41449 sha256=c2dd11e3de9f8a894c85c87fd21fc66a48616c2ed7474f7ca512617f84a10fc9\n","  Stored in directory: /root/.cache/pip/wheels/22/57/d7/b1d6bb0938c60d5dc3ed91785a8fd345cfd9fe32828bf13fd3\n","Successfully built multipy\n","Installing collected packages: multipy\n","Successfully installed multipy-0.16\n"],"name":"stdout"}]},{"cell_type":"markdown","metadata":{"id":"47FfSpiWbLNV"},"source":["## Methods for controlling the FWER\r\n","\r\n","* Bonferroni correction\r\n","* Šidák correction \r\n","* Hochberg's procedure \r\n","* Holm-Bonferroni procedure \r\n","* Permutation tests \r\n","* Random field theory (RFT) based approaches \r\n","\r\n","## Quick example"]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"gmPNs1AsbgC3","executionInfo":{"status":"ok","timestamp":1610435554553,"user_tz":-60,"elapsed":650,"user":{"displayName":"sebastian heucke","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GgU7r1bq5bIRNqaoCYdLPidtmcPCdoxFiu0JStl=s64","userId":"03230349622000046876"}},"outputId":"4d11161e-caeb-4e51-b418-4fcb0eea0415"},"source":["from multipy.data import neuhaus\r\n","from multipy.fwer import sidak\r\n","\r\n","pvals = neuhaus()\r\n","significant_pvals = sidak(pvals, alpha=0.05)\r\n","print(list(zip(['{:.4f}'.format(p) for p in pvals], significant_pvals)))"],"execution_count":4,"outputs":[{"output_type":"stream","text":["[('0.0001', True), ('0.0004', True), ('0.0019', True), ('0.0095', False), ('0.0201', False), ('0.0278', False), ('0.0298', False), ('0.0344', False), ('0.0459', False), ('0.3240', False), ('0.4262', False), ('0.5719', False), ('0.6528', False), ('0.7590', False), ('1.0000', False)]\n"],"name":"stdout"}]},{"cell_type":"markdown","metadata":{"id":"yBy2tuGUcJ5a"},"source":["## Multiple hypothesis testing problem in Bioinformatics\r\n","\r\n","What is multiple testing problem?\r\n","\r\n","* In transcriptomics experiments, we analyze thousands of genes to identify significant differentially expressed genes (DEGs) between two groups (eg. normal vs. treated) which could alter biological mechanisms of a species in response to particular treatment\r\n","* In order to identify significant DEGs in terms of fold changes between two groups, it is necessary to apply multiple hypothesis testing (analysis of hypothesis across thousands of genes simultaneously). Each individual hypothesis testing for a single gene involves whether mean of two groups are different or not.\r\n","* In multiple hypothesis testing two kinds of errors that need to consider:\r\n","Type I error (False positive or chance of making mistake): Null hypothesis is rejected when it is fail to reject (Note: Type I error is also referred as P-value)\r\n","Type II error (False negative): Null hypothesis is fail to reject when it is reject\r\n","* For example, if we want to analyze 10000 genes to identify genes with significant fold changes between two gropus, the probability of getting significant results just due to by chance is 500 (5*10000/100) at a significance level of 0.05 (5%) i.e 500 genes identified as false positive. The number of getting false positives will be more as we test large number of genes (generally in between 20K to 40K in human and plant species). Interpreting false positive results are very harmful as it will lead to wrong conclusions and therefore it is essential to control false positive discovery.\r\n","\r\n","How to minimize false discoveries?\r\n","\r\n","* To control false discoveries from multiple hypothesis testing, it is imperative adjust significance level (α) to reduce the probability of getting Type I error (family-wise error rate or FWER)\r\n","* Family-wise error rate (FWER) is defined as a probability of getting at least one significant result (false discovery) just by chance during multiple testing.\r\n","\r\n","For example, If we perform 50 multiple tests at α=0.05, the probability of getting at least one significant result by chance is ~ 92%\r\n","\r\n","P\r\n","(\r\n","at least one significant\r\n",")\r\n","=\r\n","1\r\n","−\r\n","P\r\n","(\r\n","no significant)\r\n","\r\n","=\r\n","1\r\n","−\r\n","(\r\n","1\r\n","−\r\n","0.05\r\n",")\r\n","50\r\n","∼\r\n","0.92\r\n","\r\n","In transcriptomics experiments, we analyze thousands of genes and therefore, probability of getting false positive results will be much higher.\r\n","\r\n","* FWER can be controlled by using Bonferroni correction where significance level (α) is adjusted to (α/n).\r\n","For example, If we perform 50 multiple tests at α=0.05, the probability of getting at least one significant result by chance is ~ 4%\r\n","\r\n","The new adjusted α = 0.05/50 = 0.001\r\n","\r\n","P\r\n","(\r\n","at least one significant\r\n",")\r\n","=\r\n","1\r\n","−\r\n","P\r\n","(\r\n","no significant)\r\n","\r\n","=\r\n","1\r\n","−\r\n","(\r\n","1\r\n","−\r\n","0.001\r\n",")\r\n","50\r\n","∼\r\n","0.04\r\n","\r\n","Caution: Bonferroni correction is a highly conservative method. It can decrease the rate of false positives, but also extremely increases false negatives.\r\n","\r\n","* Another approach to control the false discoveries from multiple hypothesis testing is to control false discovery rate (FDR). FDR is defined as the proportion of false positives among the significant results. Benjamini and Hochberg (1995) technique can be used to control FDR at a specific significance level. Benjamini and Hochberg is more suitable and recommended method for correcting P-values than Bonferroni correction in gene expression analysis.\r\n","\r\n","When not to use multiple hypothesis testing corrections?\r\n","\r\n","P-value corrections methods for multiple hypothesis testing can increase the number of false negatives depending on the type of methodology used. If false negative observations are very important or expensive, you should avoid correcting P-values during multiple hypothesis testing.\r\n","\r\n","Below graph represents the probability of getting at least one significant result by chance with different numbers tests at α = 0.05 and α = 0.01,\r\n"]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"70nU8BHwenz1","executionInfo":{"status":"ok","timestamp":1610436216353,"user_tz":-60,"elapsed":655,"user":{"displayName":"sebastian heucke","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GgU7r1bq5bIRNqaoCYdLPidtmcPCdoxFiu0JStl=s64","userId":"03230349622000046876"}},"outputId":"8f60135d-34d1-4233-8542-c1dadb9616ea"},"source":["# I am using Python3\r\n","# load required packages\r\n","# make sure you have installed required packages\r\n","from scipy.stats import norm\r\n","import numpy as np\r\n","from statsmodels.stats.multitest import multipletests\r\n","\r\n","# generate 500 random numbers (observations) with mean 0 and standard deviation 1 for \r\n","# multiple hypothesis testing\r\n","rand_num = norm.rvs(loc=0, scale=1, size=500)\r\n","# now, get P-values for these observations using one-sided test\r\n","# Alternate hypothesis= observation value is larger than 0\r\n","# All observations should fail to reject null hypothesis and any obtained significant \r\n","# P-value is just by a chance. We should expect ~25 (5*(500/100) significant P-value \r\n","# (at 0.05 significance level) by a chance\r\n","pval = 1-norm.cdf(rand_num) # right-sided test\r\n","# count P-values < 0.05\r\n","# output: 25\r\n","print(len(pval[np.where(pval<0.05)]))\r\n","\r\n","# Bonferroni correction\r\n","# Get Bonferroni corrected P-value, which is 0.0001\r\n","bf_p = 0.05/500 \r\n","# count P-values < bf_p\r\n","# output: 0\r\n","print(len(pval[np.where(pval<bf_p)]))\r\n","\r\n","# Benjamini and Hochberg FDR at alpha=0.05\r\n","y=multipletests(pvals=pval, alpha=0.05, method=\"fdr_bh\")\r\n","# output: 0\r\n","print(len(y[1][np.where(y[1]<0.05)]))  # y[1] returns corrected P-vals (array)"],"execution_count":6,"outputs":[{"output_type":"stream","text":["21\n","0\n","0\n"],"name":"stdout"},{"output_type":"stream","text":["/usr/local/lib/python3.6/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.\n","  import pandas.util.testing as tm\n"],"name":"stderr"}]},{"cell_type":"markdown","metadata":{"id":"gIuXemqKeoNd"},"source":["If you repeat above code, you should get significant P-values close to 25 which is also referred as Type I error rate (false positives) and calculated as (25/500=0.05). If you notice, the Type I error rate is equal to P-value significance level (0.05). After Bonferroni and Benjamini/Hochberg correction, we have obtained 0 false positives (Type I error rate=0/500=0).\r\n","\r\n","Bonferroni correction to control FWER: False positives = 0\r\n","Benjamini and Hochberg to control FDR: False positives = 0"]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"0bOpQC6resDH","executionInfo":{"status":"ok","timestamp":1610436243508,"user_tz":-60,"elapsed":671,"user":{"displayName":"sebastian heucke","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GgU7r1bq5bIRNqaoCYdLPidtmcPCdoxFiu0JStl=s64","userId":"03230349622000046876"}},"outputId":"9cf831f8-dab6-4720-9d3b-f29422c09c13"},"source":["# Now, generate 500 random numbers (observations) with mean 3 and standard deviation 1 for \r\n","# multiple hypothesis testing\r\n","rand_num = norm.rvs(loc=3, scale=1, size=500)\r\n","# now, get P-values for these observations using one-sided test\r\n","# Alternate hypothesis= observation value is larger than 0\r\n","# All observations should reject null hypothesis (as we are performing cdf with mean=0) \r\n","# and any obtained non-significant P-value is just by a chance\r\n","pval = 1-norm.cdf(rand_num) # right-sided test\r\n","# count P-values < 0.05\r\n","# output: 460\r\n","print(len(pval[np.where(pval<0.05)]))\r\n","\r\n","# Bonferroni correction\r\n","# Get Bonferroni corrected P-value, which is 0.0001\r\n","bf_p = 0.05/500 \r\n","# count P-values < bf_p\r\n","# output: 126\r\n","print(len(pval[np.where(pval<bf_p)]))\r\n","\r\n","# Benjamini and Hochberg FDR at alpha=0.05\r\n","y=multipletests(pvals=pval, alpha=0.05, method=\"fdr_bh\")\r\n","# output: 450\r\n","print(len(y[1][np.where(y[1]<0.05)]))  # y[1] returns corrected P-vals (array)"],"execution_count":7,"outputs":[{"output_type":"stream","text":["449\n","105\n","445\n"],"name":"stdout"}]},{"cell_type":"markdown","metadata":{"id":"1p9lNwG_eu1o"},"source":["If you repeat above code, you should get significant P-values close to 460. The Type II error rate (false negative) is 40/500 = 0.08. After Bonferroni correction, the rate of false negatives is tremendously increased (374/500=0.74), whereas with Benjamini/Hochberg correction, the rate of false negatives is 50/500 = 0.1.\r\n","\r\n","Bonferroni correction to control FWER: False negatives = 374\r\n","Benjamini and Hochberg to control FDR: False negatives = 50"]},{"cell_type":"code","metadata":{"id":"_Xrl2FM1eyZo"},"source":[""],"execution_count":null,"outputs":[]}]}